test/e2e/network/netpol/k8s_util.go

```diff
@@ -0,0 +1,363 @@
+func (k *Kubernetes) InitializeCluster(model *Model) error {
+	var createdPods []*v1.Pod
+	for _, ns := range model.Namespaces {
+		_, err := k.CreateOrUpdateNamespace(ns.Spec())
+		if err != nil {
+			return err
+		}
+
+		for _, pod := range ns.Pods {
+			framework.Logf("creating/updating pod %s/%s", ns.Name, pod.Name)
+
+			kubePod, err := k.CreateOrUpdatePod(pod.KubePod())
+			if err != nil {
+				return err
+			}
+			createdPods = append(createdPods, kubePod)
+
+			_, err = k.CreateOrUpdateService(pod.Service())
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, podString := range model.AllPodStrings() {
+		k8sPod, err := k.GetPodFromCache(podString.Namespace(), podString.PodName())
+		if err != nil {
+			return err
+		}
+		if k8sPod == nil {
+			return errors.Errorf("unable to find pod in ns %s with key/val pod=%s", podString.Namespace(), podString.PodName())
+		}
+		err = e2epod.WaitForPodNameRunningInNamespace(k.ClientSet, k8sPod.Name, k8sPod.Namespace)
+		if err != nil {
+			return errors.Wrapf(err, "unable to wait for pod %s/%s", podString.Namespace(), podString.PodName())
+		}
+	}
+
+	for _, createdPod := range createdPods {
+		err := e2epod.WaitForPodRunningInNamespace(k.ClientSet, createdPod)
+		if err != nil {
+			return errors.Wrapf(err, "unable to wait for pod %s/%s", createdPod.Namespace, createdPod.Name)
+		}
+	}
+
+	return nil
+}
+
+// GetPodFromCache returns a pod with the matching namespace and name
+func (k *Kubernetes) GetPodFromCache(ns string, name string) (*v1.Pod, error) {
+	pods, err := k.getPodsUncached(ns, "pod", name)
+	if err != nil {
+		return nil, err
+	}
+	if len(pods) == 0 {
+		return nil, nil
+	}
+	return &pods[0], nil
+}
+
+func (k *Kubernetes) getPodsUncached(ns string, key string, val string) ([]v1.Pod, error) {
+	v1PodList, err := k.ClientSet.CoreV1().Pods(ns).List(context.TODO(), metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%v=%v", key, val),
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to list Pods in ns %s with key/val %s=%s", ns, key, val)
+	}
+	return v1PodList.Items, nil
+}
+
+// GetPodsFromCacheByKeyVal returns an array of all Pods in the given namespace having a k/v label pair.
+func (k *Kubernetes) GetPodsFromCacheByKeyVal(ns string, key string, val string) ([]v1.Pod, error) {
+	k.mutex.Lock()
+	p, ok := k.podCache[fmt.Sprintf("%v_%v_%v", ns, key, val)]
+	k.mutex.Unlock()
+	if ok {
+		return p, nil
+	}
+
+	v1PodList, err := k.getPodsUncached(ns, key, val)
+	if err != nil {
+		return nil, err
+	}
+
+	k.mutex.Lock()
+	k.podCache[fmt.Sprintf("%v_%v_%v", ns, key, val)] = v1PodList
+	k.mutex.Unlock()
+
+	return v1PodList, nil
+}
+
+// GetPod gets a pod by namespace and name
+func (k *Kubernetes) GetPod(ns string, name string) (*v1.Pod, error) {
+	kubePod, err := k.ClientSet.CoreV1().Pods(ns).Get(context.TODO(), name, metav1.GetOptions{})
+	return kubePod, errors.Wrapf(err, "unable to get pod %s/%s", ns, name)
```

Comment:
TIL this return nil if err=nil, 
https://play.golang.org/p/qTlfACEHHD2
maybe making it explicit is easier to read for others
if err !=nil {
   return nil,errors.Wrapf 
 }

return k, nil

Reply:
👍